# Dockerfiles for bioinformatics tools


| Tool Name   | Reference                                                                                     | Version List       | 
|-------------|-----------------------------------------------------------------------------------------------|--------------------|
| nanopolish        | [A complete bacterial genome assembled de novo using only nanopore sequencing data](https://www.nature.com/articles/nmeth.3444) | ![nanopolish Version](https://img.shields.io/badge/nanopolish-0.14.0-blue) ![nanopolish Version](https://img.shields.io/badge/nanopolish-0.13.3-blue) |
| DRUMMER        |  |  |
| jaffal        |  |  |
| autodock vina        |  |  |
| hamr        | [HAMR:high-throughput annotation of modified ribonucleotides](https://rnajournal.cshlp.org/content/19/12/1684) | ![hamr Version](https://img.shields.io/badge/hamr-0a04208eae481137e0081dd90b9c1a40bc49c9f3-blue) |
| rnasieve        |  |  |
| lafite        |  |  |
| sicelore        | [High throughput error corrected Nanopore single cell transcriptome sequencing](https://doi.org/10.1038/s41467-020-17800-6) | ![sicelore Version](https://img.shields.io/badge/sicelore-2.0-blue) |
| slow5tools        | [Fast nanopore sequencing data analysis with SLOW5](https://www.nature.com/articles/s41587-021-01147-4) | ![slow5tools Version](https://img.shields.io/badge/slow5tools-1.3.0-blue) ![slow5tools Version](https://img.shields.io/badge/slow5tools-1.2.0-blue) ![slow5tools Version](https://img.shields.io/badge/slow5tools-1.1.0-blue) ![slow5tools Version](https://img.shields.io/badge/slow5tools-1.0.0-blue) |
| psinanopore        |  |  |
| deconvseq        |  |  |
| modtect        |  |  |
| stringtie        | [Improved transcriptome assembly using a hybrid of long and short reads with StringTie](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009730) | ![stringtie Version](https://img.shields.io/badge/stringtie-3.0.0-blue) ![stringtie Version](https://img.shields.io/badge/stringtie-2.2.3-blue) ![stringtie Version](https://img.shields.io/badge/stringtie-2.2.2-blue) ![stringtie Version](https://img.shields.io/badge/stringtie-2.2.1-blue) ![stringtie Version](https://img.shields.io/badge/stringtie-2.2.0-blue) ![stringtie Version](https://img.shields.io/badge/stringtie-2.1.7-blue) ![stringtie Version](https://img.shields.io/badge/stringtie-2.1.6-blue) ![stringtie Version](https://img.shields.io/badge/stringtie-2.1.5-blue) ![stringtie Version](https://img.shields.io/badge/stringtie-2.1.4-blue) ![stringtie Version](https://img.shields.io/badge/stringtie-2.1.2-blue) ![stringtie Version](https://img.shields.io/badge/stringtie-2.0-blue) ![stringtie Version](https://img.shields.io/badge/stringtie-1.3.6-blue) |
| minimap2        | [Minimap2:pairwise alignment for nucleotide sequences](https://academic.oup.com/bioinformatics/article/37/23/4572/6384570) | ![minimap2 Version](https://img.shields.io/badge/minimap2-2.28-blue) ![minimap2 Version](https://img.shields.io/badge/minimap2-2.27-blue) ![minimap2 Version](https://img.shields.io/badge/minimap2-2.26-blue) |
| gloritools        |  |  |
| bayesprism        |  |  |
| differr        |  |  |
| scaden        | [Deep learningâ€“based cell composition analysis from tissue expression profiles](https://www.science.org/doi/10.1126/sciadv.aba2619) | ![scaden Version](https://img.shields.io/badge/scaden-1.1.2-blue) ![scaden Version](https://img.shields.io/badge/scaden-1.1.1-blue) ![scaden Version](https://img.shields.io/badge/scaden-1.1.0-blue) ![scaden Version](https://img.shields.io/badge/scaden-1.0.2-blue) ![scaden Version](https://img.shields.io/badge/scaden-1.0.1-blue) ![scaden Version](https://img.shields.io/badge/scaden-1.0.0-blue) |
| genion        |  |  |
| bambu        | [Context-aware transcript quantification from long-read RNA-seq data with Bambu](https://www.nature.com/articles/s41592-023-01908-w) | ![bambu Version](https://img.shields.io/badge/bambu-3.8.3-blue) ![bambu Version](https://img.shields.io/badge/bambu-3.4.0-blue) ![bambu Version](https://img.shields.io/badge/bambu-3.2.4-blue) ![bambu Version](https://img.shields.io/badge/bambu-3.0.8-blue) ![bambu Version](https://img.shields.io/badge/bambu-3.0.6-blue) ![bambu Version](https://img.shields.io/badge/bambu-3.0.5-blue) ![bambu Version](https://img.shields.io/badge/bambu-3.0.1-blue) ![bambu Version](https://img.shields.io/badge/bambu-2.0.6-blue) ![bambu Version](https://img.shields.io/badge/bambu-2.0.3-blue) ![bambu Version](https://img.shields.io/badge/bambu-2.0.0-blue) ![bambu Version](https://img.shields.io/badge/bambu-1.2.0-blue) ![bambu Version](https://img.shields.io/badge/bambu-1.0.2-blue) ![bambu Version](https://img.shields.io/badge/bambu-1.0.0-blue) |
| tape        | [Deep autoencoder for interpretable tissue-adaptive deconvolution and cell-type-specific gene analysis](https://www.nature.com/articles/s41467-022-34550-9) | ![tape Version](https://img.shields.io/badge/tape-1.1.2-blue) |
| deeptools        |  |  |
| aeron        | [AERON:Transcript quantification and gene-fusion detection using long reads](https://www.biorxiv.org/content/10.1101/2020.01.27.921338v1.full) | ![aeron Version](https://img.shields.io/badge/aeron-a6e7d589e3feeb22b5374b455a1a677e3bb2edfa-blue) |
| bisque        |  |  |
| lafite        |  |  |
| epinano        |  |  |
| genekitr        |  |  |
| music        |  |  |
| meeko        |  |  |
| eligos2        |  |  |
| nanocount        |  |  |
| flair        | [Detecting haplotype-specific transcript variation in long reads with FLAIR2](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-024-03301-y) | ![flair Version](https://img.shields.io/badge/flair-2.0.0-blue) ![flair Version](https://img.shields.io/badge/flair-1.7.0-blue) ![flair Version](https://img.shields.io/badge/flair-1.6.4-blue) ![flair Version](https://img.shields.io/badge/flair-1.6.3-blue) ![flair Version](https://img.shields.io/badge/flair-1.5-blue) ![flair Version](https://img.shields.io/badge/flair-1.4-blue) |
| longgf        |  |  |
| mgltools        |  |  |
| ensembldb        |  |  |
| f5c        |  |  |
| biopython        | [Biopython:freely available Python tools for computational molecular biology and bioinformatics](http://dx.doi.org/10.1093/bioinformatics/btp163) | ![biopython Version](https://img.shields.io/badge/biopython-1.85-blue) ![biopython Version](https://img.shields.io/badge/biopython-1.84-blue) ![biopython Version](https://img.shields.io/badge/biopython-1.83-blue) ![biopython Version](https://img.shields.io/badge/biopython-1.82-blue) ![biopython Version](https://img.shields.io/badge/biopython-1.76-blue) ![biopython Version](https://img.shields.io/badge/biopython-1.69-blue) ![biopython Version](https://img.shields.io/badge/biopython-1.67-blue) |
| genomicfeatures        |  |  |

